src/node_http2.h

```diff
@@ -0,0 +1,572 @@
+    return *stream_buf_;
+  }
+
+ private:
+  StreamBase* stream_;
+  StreamResource::Callback<StreamResource::AllocCb> prev_alloc_cb_;
+  StreamResource::Callback<StreamResource::ReadCb> prev_read_cb_;
+  padding_strategy_type padding_strategy_ = PADDING_STRATEGY_NONE;
+  MaybeStackBuffer<char, kAllocBufferSize> stream_buf_;
+};
+
+class ExternalHeader :
+    public String::ExternalOneByteStringResource {
+ public:
+  explicit ExternalHeader(nghttp2_rcbuf* buf)
+     : buf_(buf), vec_(nghttp2_rcbuf_get_buf(buf)) {
+  }
+
+  ~ExternalHeader() override {
+    nghttp2_rcbuf_decref(buf_);
+    buf_ = nullptr;
+  }
+
+  const char* data() const override {
+    return const_cast<const char*>(reinterpret_cast<char*>(vec_.base));
+  }
+
+  size_t length() const override {
+    return vec_.len;
+  }
+
+  static Local<String> New(Isolate* isolate, nghttp2_rcbuf* buf) {
+    EscapableHandleScope scope(isolate);
+    nghttp2_vec vec = nghttp2_rcbuf_get_buf(buf);
+    if (vec.len == 0) {
+      nghttp2_rcbuf_decref(buf);
+      return scope.Escape(String::Empty(isolate));
+    }
+
+    ExternalHeader* h_str = new ExternalHeader(buf);
+    MaybeLocal<String> str = String::NewExternalOneByte(isolate, h_str);
+    isolate->AdjustAmountOfExternalAllocatedMemory(vec.len);
+
+    if (str.IsEmpty()) {
+      delete h_str;
+      return scope.Escape(String::Empty(isolate));
+    }
+
+    return scope.Escape(str.ToLocalChecked());
+  }
+
+ private:
+  nghttp2_rcbuf* buf_;
+  nghttp2_vec vec_;
+};
+
+class Headers {
+ public:
+  Headers(Isolate* isolate, Local<Context> context, Local<Array> headers) {
+    headers_.AllocateSufficientStorage(headers->Length());
+    Local<Value> item;
+    Local<Array> header;
+
+    for (size_t n = 0; n < headers->Length(); n++) {
+      item = headers->Get(context, n).ToLocalChecked();
+      CHECK(item->IsArray());
+      header = item.As<Array>();
+      Local<Value> key = header->Get(context, 0).ToLocalChecked();
+      Local<Value> value = header->Get(context, 1).ToLocalChecked();
+      CHECK(key->IsString());
+      CHECK(value->IsString());
+      size_t keylen = StringBytes::StorageSize(isolate, key, ASCII);
+      size_t valuelen = StringBytes::StorageSize(isolate, value, ASCII);
+      headers_[n].flags = NGHTTP2_NV_FLAG_NONE;
+      Local<Value> flag = header->Get(context, 2).ToLocalChecked();
+      if (flag->BooleanValue(context).ToChecked())
+        headers_[n].flags |= NGHTTP2_NV_FLAG_NO_INDEX;
+      uint8_t* buf = Malloc<uint8_t>(keylen + valuelen);
+      headers_[n].name = buf;
+      headers_[n].value = buf + keylen;
+      headers_[n].namelen =
+          StringBytes::Write(isolate,
+                            reinterpret_cast<char*>(headers_[n].name),
+                            keylen, key, ASCII);
+      headers_[n].valuelen =
+          StringBytes::Write(isolate,
+                            reinterpret_cast<char*>(headers_[n].value),
+                            valuelen, value, ASCII);
+    }
+  }
```

Comment:
This could probably be DRYed up with `CopyHeaders()` a lot, can you leave a TODO for that or something (you can assign it to me if you like)?